src/ProtoRoute/Message.hs

```diff
@@ -1,46 +1,154 @@
     { messageName :: MsgName
-    , messageFields :: [(FieldName, FieldValue)]
-    } deriving (Show)
+    , messageFields :: [MessageField]
+    } deriving (Show, Eq)
+
+-- Structure of a msg field, which has custom values
+data MessageField = MessageField
+    { messageFieldName :: FieldName
+    , messageFieldValue :: FieldValue
+    } deriving (Show, Eq)
+
+-- Validation errors can be incomplete required fields, wrong fields types,
+-- wrong rules or mismatching field names
+data ValidationError = Incomplete String |
+                       WrongName  String |
+                       WrongType  String |
+                       Invalid    String deriving (Show, Eq)
+
+incomplete :: ValidationError
+incomplete = Incomplete "Required field not filled"
+
+incorrectFieldNames :: ValidationError
+incorrectFieldNames = WrongName "Wrong field names"
+
+incorrectMsgNames :: ValidationError
+incorrectMsgNames = WrongName "Message names do not match"
+
+incorrectTypes :: ValidationError
+incorrectTypes = WrongType "There are mismatching types"
 
-data Field = Field
-    { fieldName :: FieldName
-    , fieldValue :: FieldValue
-    } deriving (Show)
+invalid :: ValidationError
+invalid = Invalid "Your message is invalid"
 
+-- Shows value of a message field
 fromFV :: FieldValue -> String
 fromFV (FText txt) = fromTV txt
 fromFV (FInt num)  = fromTV num
 fromFV (FMsg msg)  = fromTV msg
 
+-- Shows value inside FieldValue constructor
 fromTV :: (Show a) => TValue a -> String
 fromTV (Req val)  = show val
 fromTV (Opt txt)  = show txt
 fromTV (Rep list) = show list
 fromTV Error = "Error"
 
+-- Checks if rules in msg match those in corresponding schema
+correct :: TValue a -> SchemaRule -> Bool
+correct (Req _) SReq = True
+correct (Opt _) SOpt = True
+correct (Rep _) SRep = True
+correct _ _ = False
+
+-- Checks if types are correct
+okTypes :: FieldValue -> SchemaRule -> SchemaType -> Bool
+okTypes (FText tvA) rA fTA = fTA == SText && correct tvA rA
+okTypes (FInt tvB) rB fTB  = fTB == SInt && correct tvB rB
+okTypes (FMsg tvC) rC fTC  = fTC == SMsg && correct tvC rC
+
+-- Verifies if there exists a validation error
+verify :: Bool -> ValidationError -> Either ValidationError ()
+verify True someError = Left someError
+verify False _ = Right ()
+
+-- Checks if a required field is filled
+reqValThere :: SchemaField -> MessageField -> Bool
+reqValThere (SchemaField SReq _ sfn) mf =
+    case mf of
+        (MessageField sfn _) -> True
+reqValThere _ _ = False
+
+-- Validates single field, checking msg attempt against schema
+validateField :: MessageField -> SchemaField -> Either ValidationError ()
+validateField (MessageField mfn mfv) (SchemaField sfr sft sfn) =
+    case wrongFieldNames of
+        Left se  -> Left se
+        Right () ->
+          case wrongType of
+              Left se  -> Left se
+              Right () ->
+                case noReqVal of
+                    Left se  -> Left se
+                    Right () -> Right ()
```

Comment:
Use `Either` as monad here as well. You will notice that code becomes much cleaner.